File: typescript/tests/verification/smt-script-golden.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place, environmentPlace } from '../../src/core/place.js';
import type { Place } from '../../src/core/place.js';
import { one, atLeast } from '../../src/core/in.js';
import { outPlace, andPlaces } from '../../src/core/out.js';
import { matchSpec, matchKey } from '../../src/core/match-spec.js';
import { nameId } from '../../src/core/name.js';
import { MarkingState } from '../../src/verification/marking-state.js';
import { flatten } from '../../src/verification/encoding/net-flattener.js';
import type { FlatNet } from '../../src/verification/encoding/flat-net.js';
import { IncidenceMatrix } from '../../src/verification/encoding/incidence-matrix.js';
import {
  canonicalInvariantOrder, computePInvariants, computePSemiflows, validateInvariantsExact,
} from '../../src/verification/invariant/p-invariant-computer.js';
import { alwaysAvailable, bounded, ignore } from '../../src/verification/analysis/environment-analysis-mode.js';
import {
  branchPlaceBound, deadlockFree, mutualExclusion, placeBound, unreachable,
} from '../../src/verification/smt-property.js';
import { encode } from '../../src/verification/z3/smt-encoder.js';
import { vcScript } from '../../src/verification/z3/certificate-checker.js';
import { buildColouredPlan, encodeColoured } from '../../src/verification/z3/name-coloured-encoder.js';

/**
 * VER-013 AC1: the scripts this verifier sends to z3 are byte-identical to the Rust
 * reference. The goldens under `tests/fixtures/smt-golden/` were written by the Rust
 * verifier (`LIBPETRI_SMT_DUMP`) for the nets rebuilt here; a diff is a parity
 * finding in whichever emitter drifted, never a reason to edit the golden.
 *
 * No solver is needed: the encoders are pure text.
 */
const GOLDEN_DIR = join(dirname(fileURLToPath(import.meta.url)), '..', 'fixtures', 'smt-golden');

function golden(name: string): string {
  return readFileSync(join(GOLDEN_DIR, name), 'utf8');
}

/** The certificate block a golden certificate script carries (its third block). */
function certificateOf(script: string): string {
  const start = script.indexOf('(define-fun');
  const end = script.indexOf('\n\n(declare-const');
  return script.slice(start, end);
}

/** The invariants exactly as the verifier hands them to the encoders. */
function encoderInvariants(flat: FlatNet, m0: MarkingState) {
  const matrix = IncidenceMatrix.from(flat);
  return canonicalInvariantOrder(validateInvariantsExact(matrix, computePInvariants(matrix, flat, m0), flat, m0).valid);
}

function validatedSemiflows(flat: FlatNet, m0: MarkingState) {
  const matrix = IncidenceMatrix.from(flat);
  return validateInvariantsExact(matrix, computePSemiflows(matrix, flat, m0), flat, m0).valid;
}

describe('SMT script parity with the Rust reference (VER-013 AC1)', () => {
  it('chain: p0(1) -> p1, placeBound(p1, 0)', () => {
    const p0 = place('p0');
    const p1 = place('p1');
    const t = Transition.builder('t').inputs(one(p0)).outputs(outPlace(p1)).build();
    const flat = flatten(PetriNet.builder('chain').transitions(t).build(), new Set(), ignore());
    const m0 = MarkingState.builder().tokens(p0, 1).build();
    const encoding = encode(flat, m0, placeBound(p1, 0), encoderInvariants(flat, m0), new Set(), true);
    expect(encoding.smt2).toBe(golden('chain-horn.smt2'));
  });

  // arcs: inhibitor + read + reset + at-least + sink + env injection
  const A = place('A');
  const B = place('B');
  const C = place('C');
  const R = place('R');
  const S = place('S');
  const E = environmentPlace('E');

  function arcsNet(): PetriNet {
    const t1 = Transition.builder('t1')
      .inputs(one(A), one(E.place))
      .inhibitors(B)
      .reads(C)
      .resets(R)
      .outputs(outPlace(S))
      .build();
    const t2 = Transition.builder('t2').inputs(atLeast(1, B)).outputs(andPlaces(A, R)).build();
    return PetriNet.builder('arcs').transitions(t1, t2).build();
  }

  it('arcs: deadlockFree with a sink and alwaysAvailable injection', () => {
    const flat = flatten(arcsNet(), new Set([E]), alwaysAvailable());
    const m0 = MarkingState.builder().tokens(A, 1).tokens(C, 1).tokens(B, 2).build();
    const invariants = encoderInvariants(flat, m0);
    const sinks = new Set<Place<any>>([S]);
    expect(encode(flat, m0, deadlockFree(), invariants, sinks, true).smt2).toBe(golden('arcs-horn.smt2'));
    const cert = golden('arcs-certificate.smt2');
    expect(vcScript(certificateOf(cert), flat, m0, deadlockFree(), sinks, invariants)).toBe(cert);
  });

  it('arcs: unreachable with bounded(2) injection', () => {
    const flat = flatten(arcsNet(), new Set([E]), bounded(2));
    const m0 = MarkingState.builder().tokens(A, 1).build();
    const invariants = encoderInvariants(flat, m0);
    const property = unreachable(new Set([S, A]));
    expect(encode(flat, m0, property, invariants, new Set(), true).smt2).toBe(golden('unreach-horn.smt2'));
    const cert = golden('unreach-certificate.smt2');
    expect(vcScript(certificateOf(cert), flat, m0, property, new Set(), invariants)).toBe(cert);
  });

  it('mutex: mutual exclusion on a one-token cycle', () => {
    const x = place('X');
    const y = place('Y');
    const xy = Transition.builder('XtoY').inputs(one(x)).outputs(outPlace(y)).build();
    const yx = Transition.builder('YtoX').inputs(one(y)).outputs(outPlace(x)).build();
    const flat = flatten(PetriNet.builder('cycle').transitions(xy, yx).build(), new Set(), ignore());
    const m0 = MarkingState.builder().tokens(x, 1).build();
    const invariants = encoderInvariants(flat, m0);
    // Listed Y-first on purpose: the script orders the places by index.
    const property = mutualExclusion(y, x);
    expect(encode(flat, m0, property, invariants, new Set(), true).smt2).toBe(golden('mutex-horn.smt2'));
    const cert = golden('mutex-certificate.smt2');
    expect(vcScript(certificateOf(cert), flat, m0, property, new Set(), invariants)).toBe(cert);
  });

  it('nu scatter-gather with a declared budget: the name-coloured encoder', () => {
    const source = place('source');
    const budget = place('budget');
    const pending = place('pending');
    const a = place<string>('branchA');
    const b = place<string>('branchB');
    const merged = place<string>('merged');
    const fork = Transition.builder('fork')
      .inputs(one(source), one(budget))
      .outputs(andPlaces(a, b, pending))
      .build();
    const join = Transition.builder('join')
      .inputs(one(a), one(b), one(pending))
      .match(matchSpec(matchKey(a, (s: string) => nameId(s)), matchKey(b, (s: string) => nameId(s))))
      .outputs(andPlaces(merged, budget))
      .build();
    const net = PetriNet.builder('nu').transitions(fork, join).build();
    const flat = flatten(net, new Set(), ignore());
    const m0 = MarkingState.builder().tokens(source, 3).tokens(budget, 2).build();
    const plan = buildColouredPlan(net, flat, m0, new Set(['budget']), 'base', new Set(), validatedSemiflows(flat, m0));
    expect(plan, 'the scatter-gather net is in the coloured fragment').not.toBeNull();
    const encoding = encodeColoured(plan!, flat, m0, branchPlaceBound(budget, 2), encoderInvariants(flat, m0), new Set());
    expect(encoding).not.toBeNull();
    expect(encoding!.smt2).toBe(golden('nu-bound-horn-coloured.smt2'));
  });
});

File: java/src/test/java/org/libpetri/smt/SmtScriptGoldenTest.java
package org.libpetri.smt;

import org.libpetri.analysis.EnvironmentAnalysisMode;
import org.libpetri.analysis.FragmentMode;
import org.libpetri.analysis.MarkingState;
import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.libpetri.core.EnvironmentPlace;
import org.libpetri.core.MatchSpec;
import org.libpetri.core.NameId;
import org.libpetri.core.PetriNet;
import org.libpetri.core.Place;
import org.libpetri.core.Transition;
import org.libpetri.smt.encoding.FlatNet;
import org.libpetri.smt.encoding.NetFlattener;
import org.libpetri.smt.z3.CertificateChecker;
import org.libpetri.smt.z3.NameColouredEncoder;
import org.libpetri.smt.z3.SmtEncoder;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

/**
 * VER-013 AC1: the scripts this verifier sends to z3 are byte-identical to the Rust
 * reference. The goldens under {@code src/test/resources/smt-golden/} were written by
 * the Rust verifier ({@code LIBPETRI_SMT_DUMP}) for the nets rebuilt here; a diff is a
 * parity finding in whichever emitter drifted, never a reason to edit the golden.
 *
 * <p>No solver is needed: the encoders are pure text.
 */
class SmtScriptGoldenTest {

    private static String golden(String name) {
        try (InputStream in = SmtScriptGoldenTest.class.getResourceAsStream("/smt-golden/" + name)) {
            assertNotNull(in, "missing golden " + name);
            return new String(in.readAllBytes(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new AssertionError(e);
        }
    }

    /** The certificate block a golden certificate script carries (its third block). */
    private static String certificateOf(String script) {
        int start = script.indexOf("(define-fun");
        int end = script.indexOf("\n\n(declare-const");
        return script.substring(start, end);
    }

    private static FlatNet flatten(PetriNet net) {
        return NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());
    }

    // --- chain: p0(1) -> p1, placeBound(p1, 0) ---

    @Test
    void chain_hornScriptMatchesRust() {
        var p0 = Place.of("p0", Integer.class);
        var p1 = Place.of("p1", Integer.class);
        var t = Transition.builder("t").inputs(In.one(p0)).outputs(Out.place(p1)).build();
        var net = PetriNet.builder("chain").transitions(t).build();
        var flat = flatten(net);
        var m0 = MarkingState.builder().tokens(p0, 1).build();
        var invariants = SmtVerifier.encoderInvariants(flat, m0, false);

        var encoding = SmtEncoder.encode(flat, m0, SmtProperty.placeBound(p1, 0), invariants, Set.of(), true);
        assertEquals(golden("chain-horn.smt2"), encoding.smt2());
    }

    // --- arcs: inhibitor + read + reset + at-least + sink + env alwaysAvailable ---

    private static final Place<Integer> A = Place.of("A", Integer.class);
    private static final Place<Integer> B = Place.of("B", Integer.class);
    private static final Place<Integer> C = Place.of("C", Integer.class);
    private static final Place<Integer> R = Place.of("R", Integer.class);
    private static final Place<Integer> S = Place.of("S", Integer.class);
    private static final Place<Integer> E = Place.of("E", Integer.class);

    private static PetriNet arcsNet() {
        var t1 = Transition.builder("t1")
            .inputs(In.one(A), In.one(E))
            .inhibitors(B)
            .reads(C)
            .resets(R)
            .outputs(Out.place(S))
            .build();
        var t2 = Transition.builder("t2")
            .inputs(In.atLeast(1, B))
            .outputs(Out.and(A, R))
            .build();
        return PetriNet.builder("arcs").transitions(t1, t2).build();
    }

    /**
     * The {@code arcs-*} goldens were written by Rust for {@code deadlockFree()} with a
     * declared sink, BEFORE the [VER-002] split. Those exact bytes are now the encoding
     * of {@link SmtProperty.TerminatesAtSink} — the property that inherited the old
     * permissive predicate unchanged — so the test is retargeted rather than the goldens
     * regenerated. Nothing is lost: the arc encoding (inhibitor / read / reset / at-least)
     * and the VER-006 injection relaxation live in the shared quiescence core, which both
     * properties conjoin, and strict {@code DeadlockFree} stays pinned byte for byte by
     * the shared fixture goldens under {@code spec/verification-fixtures/scripts/}.
     */
    @Test
    void arcs_terminatesAtSinkWithSinkAndInjection_matchesRust() {
        var flat = NetFlattener.flatten(arcsNet(), Set.of(EnvironmentPlace.of(E)),
            EnvironmentAnalysisMode.alwaysAvailable());
        var m0 = MarkingState.builder().tokens(A, 1).tokens(C, 1).tokens(B, 2).build();
        var invariants = SmtVerifier.encoderInvariants(flat, m0, false);

        var encoding = SmtEncoder.encode(
            flat, m0, SmtProperty.terminatesAtSink(), invariants, Set.of(S), true);
        assertEquals(golden("arcs-horn.smt2"), encoding.smt2());

        var certGolden = golden("arcs-certificate.smt2");
        assertEquals(certGolden, CertificateChecker.vcScript(
            certificateOf(certGolden), flat, m0, SmtProperty.terminatesAtSink(), Set.of(S),
            invariants));
    }

    @Test
    void arcs_unreachableWithBoundedInjection_matchesRust() {
        var flat = NetFlattener.flatten(arcsNet(), Set.of(EnvironmentPlace.of(E)),
            EnvironmentAnalysisMode.bounded(2));
        var m0 = MarkingState.builder().tokens(A, 1).build();
        var invariants = SmtVerifier.encoderInvariants(flat, m0, false);
        var property = SmtProperty.unreachable(Set.of(S, A));

        var encoding = SmtEncoder.encode(flat, m0, property, invariants, Set.of(), true);
        assertEquals(golden("unreach-horn.smt2"), encoding.smt2());

        var certGolden = golden("unreach-certificate.smt2");
        assertEquals(certGolden, CertificateChecker.vcScript(
            certificateOf(certGolden), flat, m0, property, Set.of(), invariants));
    }

    // --- mutual exclusion on a one-token cycle ---

    @Test
    void mutex_cycle_matchesRust() {
        var x = Place.of("X", Integer.class);
        var y = Place.of("Y", Integer.class);
        var xy = Transition.builder("XtoY").inputs(In.one(x)).outputs(Out.place(y)).build();
        var yx = Transition.builder("YtoX").inputs(In.one(y)).outputs(Out.place(x)).build();
        var flat = flatten(PetriNet.builder("cycle").transitions(xy, yx).build());
        var m0 = MarkingState.builder().tokens(x, 1).build();
        var invariants = SmtVerifier.encoderInvariants(flat, m0, false);
        // Listed Y-first on purpose: the script orders the places by index.
        var property = SmtProperty.mutualExclusion(y, x);

        var encoding = SmtEncoder.encode(flat, m0, property, invariants, Set.of(), true);
        assertEquals(golden("mutex-horn.smt2"), encoding.smt2());

        var certGolden = golden("mutex-certificate.smt2");
        assertEquals(certGolden, CertificateChecker.vcScript(
            certificateOf(certGolden), flat, m0, property, Set.of(), invariants));
    }

    // --- ν scatter-gather, budget declared: the name-coloured encoder ---

    @Test
    void nuScatterGather_colouredScriptMatchesRust() {
        var source = Place.of("source", Integer.class);
        var budget = Place.of("budget", Integer.class);
        var pending = Place.of("pending", Integer.class);
        var a = Place.of("branchA", String.class);
        var b = Place.of("branchB", String.class);
        var merged = Place.of("merged", String.class);
        var fork = Transition.builder("fork")
            .inputs(In.one(source), In.one(budget))
            .outputs(Out.and(a, b, pending))
            .build();
        var join = Transition.builder("join")
            .inputs(In.one(a), In.one(b), In.one(pending))
            .match(MatchSpec.builder()
                .key(a, (String s) -> NameId.of(s))
                .key(b, (String s) -> NameId.of(s))
                .build())
            .outputs(Out.and(merged, budget))
            .build();
        var net = PetriNet.builder("nu").transitions(fork, join).build();
        var flat = flatten(net);
        var m0 = MarkingState.builder().tokens(source, 3).tokens(budget, 2).build();
        var invariants = SmtVerifier.encoderInvariants(flat, m0, false);
        var semiflows = SmtVerifier.validatedSemiflows(flat, m0);

        var plan = NameColouredEncoder.buildPlan(
            net, flat, m0, Set.of("budget"), FragmentMode.BASE, Set.of(), semiflows);
        assertNotNull(plan, "the scatter-gather net is in the coloured fragment");
        var encoding = NameColouredEncoder.encode(
            plan, flat, m0, SmtProperty.branchPlaceBound(budget, 2), invariants, Set.of());
        assertNotNull(encoding);
        assertEquals(golden("nu-bound-horn-coloured.smt2"), encoding.smt2());
    }
}
